File: swift-bridge/Sources/MapKitBridge/Address.swift
import Foundation
import MapKit

enum MKRAddressFilterModePayload: String, Codable {
    case including
    case excluding
    case includingAll
    case excludingAll
}

struct MKRAddressFilterPayload: Codable {
    var mode: MKRAddressFilterModePayload
    var options: UInt64
}

struct MKRAddressPayload: Codable {
    var fullAddress: String
    var shortAddress: String?
}

struct MKRAddressRepresentationsPayload: Codable {
    var cityName: String?
    var cityWithContext: String?
    var cityWithContextShort: String?
    var cityWithContextFull: String?
    var regionName: String?
    var regionCode: String?
    var fullAddressIncludingRegionMultiline: String?
    var fullAddressIncludingRegionSingleLine: String?
    var fullAddressExcludingRegionMultiline: String?
    var fullAddressExcludingRegionSingleLine: String?
}

@available(macOS 26.0, *)
func mkrBuildAddress(_ payload: MKRAddressPayload?) throws -> MKAddress? {
    guard let payload else { return nil }
    guard let address = MKAddress(fullAddress: payload.fullAddress, shortAddress: payload.shortAddress) else {
        throw NSError(
            domain: "mapkit-rs",
            code: -1,
            userInfo: [NSLocalizedDescriptionKey: "failed to create MKAddress"]
        )
    }
    return address
}

@available(macOS 26.0, *)
func mkrEncodeAddress(_ address: MKAddress?) -> MKRAddressPayload? {
    guard let address else { return nil }
    return MKRAddressPayload(fullAddress: address.fullAddress, shortAddress: address.shortAddress)
}

@available(macOS 26.0, *)
func mkrEncodeAddressRepresentations(
    _ addressRepresentations: MKAddressRepresentations?
) -> MKRAddressRepresentationsPayload? {
    guard let addressRepresentations else { return nil }
    return MKRAddressRepresentationsPayload(
        cityName: addressRepresentations.cityName,
        cityWithContext: addressRepresentations.cityWithContext,
        cityWithContextShort: addressRepresentations.cityWithContext(.short),
        cityWithContextFull: addressRepresentations.cityWithContext(.full),
        regionName: addressRepresentations.regionName,
        regionCode: nil,
        fullAddressIncludingRegionMultiline: addressRepresentations.fullAddress(includingRegion: true, singleLine: false),
        fullAddressIncludingRegionSingleLine: addressRepresentations.fullAddress(includingRegion: true, singleLine: true),
        fullAddressExcludingRegionMultiline: addressRepresentations.fullAddress(includingRegion: false, singleLine: false),
        fullAddressExcludingRegionSingleLine: addressRepresentations.fullAddress(includingRegion: false, singleLine: true)
    )
}

@available(macOS 15.0, *)
func mkrBuildAddressFilter(_ payload: MKRAddressFilterPayload?) throws -> MKAddressFilter? {
    guard let payload else { return nil }
    let options = MKAddressFilter.Options(rawValue: UInt(payload.options))
    switch payload.mode {
    case .including:
        return MKAddressFilter(including: options)
    case .excluding:
        return MKAddressFilter(excluding: options)
    case .includingAll:
        return .includingAll
    case .excludingAll:
        return .excludingAll
    }
}


@available(macOS 15.0, *)
func mkrEncodeAddressFilter(_ filter: MKAddressFilter?) -> MKRAddressFilterPayload? {
    guard let filter else { return nil }

    let knownOptions: [(UInt64, MKAddressFilter.Options)] = [
        (1 << 0, .country),
        (1 << 1, .administrativeArea),
        (1 << 2, .subAdministrativeArea),
        (1 << 3, .locality),
        (1 << 4, .subLocality),
        (1 << 5, .postalCode)
    ]

    var included: UInt64 = 0
    var excluded: UInt64 = 0
    for (bits, option) in knownOptions {
        if filter.includes(option) {
            included |= bits
        }
        if filter.excludes(option) {
            excluded |= bits
        }
    }

    if included == 0b11_1111 {
        return MKRAddressFilterPayload(mode: .includingAll, options: included)
    }
    if excluded == 0b11_1111 {
        return MKRAddressFilterPayload(mode: .excludingAll, options: excluded)
    }
    if included != 0 {
        return MKRAddressFilterPayload(mode: .including, options: included)
    }
    if excluded != 0 {
        return MKRAddressFilterPayload(mode: .excluding, options: excluded)
    }
    return nil
}

File: src/address.rs
use std::ops::{BitOr, BitOrAssign};

use serde::{Deserialize, Serialize};

/// Wraps `MKAddressFilterOption`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct MKAddressFilterOption(pub u64);

impl MKAddressFilterOption {
    /// Wraps `MKAddressFilterOption.country`.
    pub const COUNTRY: Self = Self(1 << 0);
    /// Wraps `MKAddressFilterOption.administrativeArea`.
    pub const ADMINISTRATIVE_AREA: Self = Self(1 << 1);
    /// Wraps `MKAddressFilterOption.subAdministrativeArea`.
    pub const SUB_ADMINISTRATIVE_AREA: Self = Self(1 << 2);
    /// Wraps `MKAddressFilterOption.locality`.
    pub const LOCALITY: Self = Self(1 << 3);
    /// Wraps `MKAddressFilterOption.subLocality`.
    pub const SUB_LOCALITY: Self = Self(1 << 4);
    /// Wraps `MKAddressFilterOption.postalCode`.
    pub const POSTAL_CODE: Self = Self(1 << 5);
    /// Wraps `MKAddressFilterOption.all`.
    pub const ALL: Self = Self(
        Self::COUNTRY.0
            | Self::ADMINISTRATIVE_AREA.0
            | Self::SUB_ADMINISTRATIVE_AREA.0
            | Self::LOCALITY.0
            | Self::SUB_LOCALITY.0
            | Self::POSTAL_CODE.0,
    );

    /// Wraps `MKAddressFilterOption.bits`.
    pub const fn bits(self) -> u64 {
        self.0
    }

    /// Wraps `MKAddressFilterOption.contains`.
    pub const fn contains(self, other: Self) -> bool {
        self.0 & other.0 == other.0
    }
}

impl BitOr for MKAddressFilterOption {
    type Output = Self;

    fn bitor(self, rhs: Self) -> Self::Output {
        Self(self.0 | rhs.0)
    }
}

impl BitOrAssign for MKAddressFilterOption {
    fn bitor_assign(&mut self, rhs: Self) {
        self.0 |= rhs.0;
    }
}

/// Wraps `MKAddressFilterMode`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum MKAddressFilterMode {
    Including,
    Excluding,
    IncludingAll,
    ExcludingAll,
}

/// Wraps `MKAddressFilter`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKAddressFilter {
    /// Wraps `MKAddressFilter.mode`.
    pub mode: MKAddressFilterMode,
    /// Wraps `MKAddressFilter.options`.
    pub options: MKAddressFilterOption,
}

impl MKAddressFilter {
    /// Wraps `MKAddressFilter.includingAll`.
    pub const fn including_all() -> Self {
        Self {
            mode: MKAddressFilterMode::IncludingAll,
            options: MKAddressFilterOption::ALL,
        }
    }

    /// Wraps `MKAddressFilter.excludingAll`.
    pub const fn excluding_all() -> Self {
        Self {
            mode: MKAddressFilterMode::ExcludingAll,
            options: MKAddressFilterOption::ALL,
        }
    }

    /// Wraps `MKAddressFilter.including`.
    pub const fn including(options: MKAddressFilterOption) -> Self {
        Self {
            mode: MKAddressFilterMode::Including,
            options,
        }
    }

    /// Wraps `MKAddressFilter.excluding`.
    pub const fn excluding(options: MKAddressFilterOption) -> Self {
        Self {
            mode: MKAddressFilterMode::Excluding,
            options,
        }
    }

    /// Wraps `MKAddressFilter.includesOptions`.
    pub const fn includes_options(&self, options: MKAddressFilterOption) -> bool {
        match self.mode {
            MKAddressFilterMode::Including => self.options.contains(options),
            MKAddressFilterMode::IncludingAll => true,
            MKAddressFilterMode::Excluding | MKAddressFilterMode::ExcludingAll => false,
        }
    }

    /// Wraps `MKAddressFilter.excludesOptions`.
    pub const fn excludes_options(&self, options: MKAddressFilterOption) -> bool {
        match self.mode {
            MKAddressFilterMode::Excluding => self.options.contains(options),
            MKAddressFilterMode::ExcludingAll => true,
            MKAddressFilterMode::Including | MKAddressFilterMode::IncludingAll => false,
        }
    }
}

/// Wraps `MKAddress`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKAddress {
    /// Wraps `MKAddress.fullAddress`.
    pub full_address: String,
    /// Wraps `MKAddress.shortAddress`.
    pub short_address: Option<String>,
}

impl MKAddress {
    /// Creates a wrapper for `MKAddress`.
    pub fn new(full_address: impl Into<String>, short_address: Option<impl Into<String>>) -> Self {
        Self {
            full_address: full_address.into(),
            short_address: short_address.map(Into::into),
        }
    }
}

/// Wraps `MKAddressRepresentationsContextStyle`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum MKAddressRepresentationsContextStyle {
    Automatic,
    Short,
    Full,
}

/// Wraps `MKAddressRepresentations`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKAddressRepresentations {
    /// Wraps `MKAddressRepresentations.cityName`.
    pub city_name: Option<String>,
    /// Wraps `MKAddressRepresentations.cityWithContext`.
    pub city_with_context: Option<String>,
    /// Wraps `MKAddressRepresentations.cityWithContextShort`.
    pub city_with_context_short: Option<String>,
    /// Wraps `MKAddressRepresentations.cityWithContextFull`.
    pub city_with_context_full: Option<String>,
    /// Wraps `MKAddressRepresentations.regionName`.
    pub region_name: Option<String>,
    /// Wraps `MKAddressRepresentations.regionCode`.
    pub region_code: Option<String>,
    /// Wraps `MKAddressRepresentations.fullAddressIncludingRegionMultiline`.
    pub full_address_including_region_multiline: Option<String>,
    /// Wraps `MKAddressRepresentations.fullAddressIncludingRegionSingleLine`.
    pub full_address_including_region_single_line: Option<String>,
    /// Wraps `MKAddressRepresentations.fullAddressExcludingRegionMultiline`.
    pub full_address_excluding_region_multiline: Option<String>,
    /// Wraps `MKAddressRepresentations.fullAddressExcludingRegionSingleLine`.
    pub full_address_excluding_region_single_line: Option<String>,
}

impl MKAddressRepresentations {
    /// Wraps `MKAddressRepresentations.fullAddress`.
    pub fn full_address(&self, including_region: bool, single_line: bool) -> Option<&str> {
        match (including_region, single_line) {
            (true, true) => self.full_address_including_region_single_line.as_deref(),
            (true, false) => self.full_address_including_region_multiline.as_deref(),
            (false, true) => self.full_address_excluding_region_single_line.as_deref(),
            (false, false) => self.full_address_excluding_region_multiline.as_deref(),
        }
    }

    /// Wraps `MKAddressRepresentations.cityWithContextUsingStyle`.
    pub fn city_with_context_using_style(
        &self,
        style: MKAddressRepresentationsContextStyle,
    ) -> Option<&str> {
        match style {
            MKAddressRepresentationsContextStyle::Automatic => self.city_with_context.as_deref(),
            MKAddressRepresentationsContextStyle::Short => self.city_with_context_short.as_deref(),
            MKAddressRepresentationsContextStyle::Full => self.city_with_context_full.as_deref(),
        }
    }
}
